Add asserts to validate that a resource exists when used

DavidPeicho opened the pull request:
Hi,

As discussed in #610, this is mostly for C/C++ users. Some questions:
* Would it be possible to display the ID of the resource that failed? I don't know what you think about implementing the `Display` trait on the `Id` struct? 
* Do you see other places needing for checks?

Thanks!

kvark commented on wgpu-core/src/device/mod.rs:
```diff
@@ -1294,6 +1294,11 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
                     binding_model::BindingResource::Buffer(ref bb) => {
+                        if cfg!(debug_assertions) {
+                            // Asserts that the resource buffer exists.
+                            assert!(buffer_guard.contains(bb.buffer), "Buffer is invalid");
```
would that be different from `debug_assert!`?

DavidPeicho replied:
Wasn't aware of `debug_assert!`, pretty neat 👍

kvark commented:
Thank you for the PR! I thought about it, and there are many many places where our assumption may not be held by a badly formed C client. We don't want this kind of check to be pervasively everywhere.
However, we can put it at the lower level. We are checking for ID to belong to a storage anyway (and panicking if it's not, like in #610 ), and we are checking the Epoch. What we don't have is a proper error message [here](https://github.com/gfx-rs/wgpu/blob/70154373f8668b88c43fcbdee7bee8a89816eb39/wgpu-core/src/hub.rs#L90).

We could add a `kind: &'static str` field to `struct Storage` to help us describing what the thing is. Then we'd do something like this:
```rust
let (ref value, storage_epoch) = self.map
  .get(index as usize)
  .unwrap_or_else(|| panic!("{}[{}] does not exist", self.kind, index));
assert_eq!(epoch, storage_epoch, "{}[{}] is no longer alive", self.kind, index);
```

DavidPeicho commented:
Sounds like a better idea yes! I will make the changes. Do you want to keep the assertion only when compiling in debug mode?

kvark commented:
> Do you want to keep the assertion only when compiling in debug mode?

In the example code I provided, it's a `panic!()` and not an assertion. Rust has bounds checks enabled in both release and debug, so the check is there anyway, we are just going to handle the case where it fails more explicitly.

DavidPeicho commented:
Indeed! I am going to fix that soon, thanks for the help

DavidPeicho commented:
Implemented your idea. It works nicely, but I feel weird to have to forward the `kind` identifier when creating a `Storage`. Wouldn't it make more sense to have something like an Associated Constant on the `TypeId` trait? I am not super familiar with Rust but it looks impossible to do it because it's not possible to specialize the trait. You anyway know more about Rust than myself but I would be glad to hear the solution if something cleaner can be achieved.

Small question oriented to Rust, how would you fix : [this "issue"](https://github.com/DavidPeicho/wgpu/blob/feature/resource-bind-assert/wgpu-core/src/hub.rs#L107). I can't use a closure here because I would end up with a mut and non-mut borrow. What's the best way to fix an error like that? I just fixed it by not using `unwrap_or_else` and so by not using any closure.

kvark commented:
> Wouldn't it make more sense to have something like an Associated Constant on the TypeId trait?

We could have it as an associated type, but it would not make the code simpler, or make it less code. So all other equals, a simpler solution (like the one you have now) is preferred.

> I also found another place where the ID could be wrong. This will happen only with pipelines I guess?

This would be wrong if you provide complete garbage in place of render/compute pass. I don't think it's something you'd want to handle specifically - there are likely many other asserts that would fire up in this case, not all of them pretty.

> I can't use a closure here because I would end up with a mut and non-mut borrow.

I actually like the `match` variant much much better. Could we turn first one into a match as well?

If you did want to work around the closure problem, you'd do something like `let kind = self.kind;` to use in it, and that would make Rust to not capture the whole `self` by the closure.

DavidPeicho commented:
> This would be wrong if you provide complete garbage in place of render/compute pass

It's an extreme case but for instance, doing:
```c
WGPUComputePassId command_pass = wgpu_command_encoder_begin_compute_pass(encoder, NULL);
wgpu_compute_pass_set_pipeline(command_pass, 0);
```

It sounds dumb like that, but it could be hidden behind more code. But I guess yes we could let it as-is, as it may be the only way to trigger it.

> I actually like the match variant much much better. Could we turn first one into a match as well?

👍 

> you'd do something like `let kind = self.kind;`

I imagine that if the closure needs a reference to an object inside `self`, it doesn't work anymore. In this case the only option would be to clone I guess? Or use something like `RefCell` if it's really needed?

This is not related to this PR but educational for me :)

kvark commented:
> It's an extreme case but for instance, doing:

Yeah, you could also pass invalid pointers to any place in the API, and we have no way of handling it. There is only so much we can do in a C API for safety. Protecting against garbage data is not something we can fix entirely, we shouldn't try too hard to address more of it.

A better focus should be on handling cases where the user is well-intentioned but accidentally misused the API, like the original case you hit.

> I imagine that if the closure needs a reference to an object inside self, it doesn't work anymore. 

No, actually, it would still work. You could do `let kind = &self.kind`. That would borrow `kind` field and not consider the whole `self` to be used.

DavidPeicho commented:
Thanks!
